### PR TITLE
refactor: transportMode and client URL options

### LINKED ISSUE #2535: devServer.sockPath
# Bug report

**What is the current behavior?**
1. Using "devserver.sockPath" will make hot updates cross domain.
2. Devserver.sockPath does not work.

**If the current behavior is a bug, please provide the steps to reproduce.**
1. Use nginx to proxy the local service. The local service uses domain name access. If you do not use "devserver.sockPath", there will be no cross domain problem. If you use it, there will be cross domain problem.
2. Devserver.sockPath itself is not working.

**What is the expected behavior?**
1. Correct hot update with devserver.sockPath.
2. Devserver.sockPath works normally.

**Other relevant information:**
webpack version:  4.43.0
Node.js version: 12.16.1
Operating System: win10
Additional tools: nginx, yarn

### PR DESCRIPTION
**Pull request by alexander-akait**
- [x] This is a **bugfix**
- [ ] This is a **feature**
- [x] This is a **code refactor**
- [ ] This is a **test update**
- [ ] This is a **docs update**
- [ ] This is a **metadata update**

### For Bugs and Features; did you add new tests?

WIP

### Motivation / Use-Case

I want to split https://github.com/webpack/webpack-dev-server/pull/3181 on multiple PRs/commits
1. We will implement (this PR) `{ options.transportMode: { server: { type: 'ws', options: { path: '/ws' } } } }` and `{ options.transportMode: { client: { type: 'ws', options: { path: '/ws' } } } }`, now we can pass options to websocket server, also should fix #2535, so you can pass
2. Move `client.host`/`client.port` to `options.transportMode.client.options`
3. Union `transportMode` and `client` options we need discussion here how better name this option, we have `server` (i.e. all top level options like `host`, `port` and etc), `websocket` server (i.e. `ws`/`sockjs` packages and options, for example `options.transportMode.server.options.path`) and `client` (i.e. websocket API/sockjs client and options for client like `client.overlay`/`client.logging`/etc), and here we have problems our options mixed between `options.transportMode.client` and top level `client` (step 2 partial fix it, but I think we can simplify it and do it better)

/cc @webpack/cli-team 

### Breaking Changes

Yes

### Additional Info

- Need fix CLI args
- Need fix examples and docs

### CODE REVIEW

**Review by evenstensberg — APPROVED** [review]
Lgtm. Nice comments on new code as well👏🏽

**Review by snitin315 — COMMENTED** [review]
Looks great! Let's fix CI 🚀

**Comment by alexander-akait** [comment]
I found better way without big changes

**Comment by alexander-akait** [comment]
Hard...

Now we have more clearly options:
- `webSocketServer` allow to set `host`/`port`/`path` (`host` and `port` is not supported by `sockjs`), it allow to run socket server even on other host and port, not only change path, by default we use `noServer` mode, but if you provide `host` or `port` it will be separate server, it is really very flexible 
- Also we have `client.host`/`client.port`/`client.path`, it change only URL to socket server, it is not change websocket server options, i.e. if you set `webSocketServer.options.path: '/custom'` it will create web socket server and will accept connection on by this path, if you set `client.path: '/custom'` it just change URL to websocket server i.e. `new WebSocket(url)` in browser. DO we really need this (maybe we have cases with proxy?)?

It's all pretty flexible but may be slightly misleading. Maybe somebody have ideas how to improve this (renaming/clear description/etc).

**Comment by snitin315** [comment]
The change looks good to me. I think we can roll it in next beta release for more feedback and see if developers files any complaints or finds it confusing.

**Comment by alexander-akait** [comment]
Maybe we don't need `client.host`/`client.port`/`client.path`, I tried to find case where it can be useful...

**Comment by alexander-akait** [comment]
@snitin315 Found cases, I think we need better name, maybe `client.webSocketHost`/etc, but still misleading, we need to find an explicit name that indicates that this does not change the web socket server host/port/path, but used only for creating *URL* to websocket server, i.e. https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/WebSocket

**Comment by alexander-akait** [comment]
I'll try to figure it out by tomorrow, I will be glad for any help, we need to finish the release as soon as possible...

**Comment by snitin315** [comment]
> @snitin315 Found cases, I think we need better name, maybe `client.webSocketHost`/etc, but still misleading, we need to find an explicit name that indicates that this does not change the web socket server host/port/path, but used only for creating _URL_ to websocket server, i.e. https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/WebSocket


Maybe `webSocketURLHost`, `webSocketURLPort` etc and we can add in the description it's only for URL. Anyway let's do it separate I can create PR for the same. Let me know.

In the meantime, I will fix the tests here.

**Comment by alexander-akait** [comment]
Yep, let's solve separately, also we need refactor our logic to inject

**Comment by alexander-akait** [comment]
I will finish PR, we need rewrite some tests + add more them, give me 30-40 minutes

File: test/fixtures/dev-server/default-config.js [hunk 1]
```diff
@@ -9,12 +9,9 @@ module.exports = {
   entry: resolve(__dirname, './foo.js'),
   devServer: {
     client: {
+      transport: 'sockjs',
```
Comment (alexander-akait) [inline test/fixtures/dev-server/default-config.js:12]: @snitin315 Again, here you do the same problem as do other developers, `client.transport` for only client, it is not create sockjs server, we need search better name for this, you should not set `sockjs`/`ws` server using this options, it is only if you want create custom web socket **client**

Comment (snitin315) [reply test/fixtures/dev-server/default-config.js:12]: I see we need configure `webSocketServer`  here.

Comment (alexander-akait) [reply test/fixtures/dev-server/default-config.js:12]: Yep, I am working on it, most of tests are good, anyway we need refactor many tests, I will create roadmap for rc

**Comment by alexander-akait** [comment]
@snitin315 @anshumanv let's finish this firstly before merge other, each rebase is harder and harder

**Comment by anshumanv** [comment]
Yep lets focus to get this in first 👍

**Review by snitin315 — APPROVED** [review]

**Review by anshumanv — APPROVED** [review]
👍

**Pull request merged**